File: src/jeeves_airframe/reward/tool_success.py
"""ToolSuccessRateReward — reward based on tool call success rate."""

from __future__ import annotations

from jeeves_airframe.trajectory.types import Step


class ToolSuccessRateReward:
    """Scores the fraction of successful tool calls in a step.

    Returns 1.0 if all tools succeeded (or no tools were called),
    0.0 if all failed, proportional otherwise.
    """

    def __init__(self, *, no_tools_score: float = 1.0):
        self._no_tools_score = no_tools_score

    @property
    def name(self) -> str:
        return "tool_success_rate"

    def score(self, step: Step) -> float:
        results = step.stage_trace.tool_results
        if not results:
            return self._no_tools_score
        return sum(1.0 for r in results if r.success) / len(results)

File: src/reward/tool_success.rs
//! ToolSuccessRateReward — fraction of successful tool calls.

use crate::reward::RewardFn;
use crate::trajectory::Step;

/// Scores the fraction of successful tool calls.
#[derive(Debug)]
pub struct ToolSuccessRateReward {
    no_tools_score: f64,
}

impl ToolSuccessRateReward {
    pub fn new(no_tools_score: f64) -> Self {
        Self { no_tools_score }
    }
}

impl RewardFn for ToolSuccessRateReward {
    fn name(&self) -> &str {
        "tool_success_rate"
    }

    fn score(&self, step: &Step) -> f64 {
        let results = &step.stage_trace.tool_results;
        if results.is_empty() {
            return self.no_tools_score;
        }
        let successes = results.iter().filter(|r| r.success).count() as f64;
        successes / results.len() as f64
    }
}
